## Programming Assignment 1 : Odyssey Villagomez - Graduate Student
#### Machine Learning (CSCI-4930/5930)

In [128]:
# Import libraries
import pandas as pd

In [129]:
# First load the dataset into pandas dataframe
training = pd.read_csv('dataset/training.csv',header=None,delimiter=',')
validation = pd.read_csv('dataset/validation.csv',header=None,delimiter=',')

In [130]:
validation.head()

0    1         2         3         4         5         6         7  8
0 -315  75   0.205128  0.230769  0.307692  0.146667  0.280000  0.240000  0
1 -243  75   0.213333  0.186667  0.293333  0.400000  0.253333  0.186667  0
2 -289  77   0.192308  0.230769  0.282051  0.220779  0.220779  0.272727  0
3 -776  118  0.271186  0.203390  0.271186  0.200000  0.200000  0.291667  1
4 -388  77   0.226190  0.238095  0.250000  0.272727  0.272727  0.207792  1

## Task : A
Print total number of samples in the validation dataset: "dataset/validation.csv"

In [131]:
print("Total number of samples: ", len(validation))

Total number of samples:  60000


## Task : B
Print two numbers in the format: [n0, n1], where
n0 represents number of class=0 (negative) samples and n1 represents number of class=1 (positive) samples in the validation dataset: "dataset/validation.csv"


In [132]:
#the class, [0,1] is in the last column: index : 8
negative_class = validation[validation[8] == 0] #filter the 0 values and assign to variable
positive_class = validation[validation[8] == 1] #filter the 0 values and assign to variable

print('[',len(negative_class), ',', len(positive_class), ']')


[ 40060 , 19940 ]


## Task : C
Print standard deviation of the second feature: "The length of shorter sequence" of the validation dataset: "dataset/validation.csv"

In [133]:
print(validation[1].std())

21.278652371559346


## Task : D
Print median (i.e., 50% percentile) of the seventh feature: "'U' frequencies of sequence 2" of the validation dataset: "dataset/validation.csv"

In [134]:
print(validation[6].median())

0.2203389999999999


## Task : E
Complete the function "confusion_matrix" partially defined that takes two arrays of target variable "y": y_actual and y_pred denoting ground truth class labels and predicted class labels for the N samples when N is the length of both the arrays. The function should return a list of 4 metrics: TN, FP, FN, TP (in this order).

In [135]:
def confusion_matrix(y_actual, y_pred):
    # the function takes two arrays of target variable "y": y_actual and y_pred
    #    denoting ground truth class labels and predicted class labels for the N samples
    #    when N is the length of both the arrays.
    # The function should return a list of 4 metrics: TN, FP, FN, TP (in this order).
    assert(len(y_actual)==len(y_pred))
    
    TP = 0
    FP = 0
    TN = 0
    FN = 0
    #@TODO
    #compare actual to predicted 
    for actual, pred in zip(y_actual, y_pred):
        if actual == 1 and pred == 1:
                TP = TP +1
        if actual == 1 and pred == 0:
                FN = FN + 1 
        if actual == 0 and pred == 0:
                TN = TN + 1 
        if actual == 0 and pred == 1:
                FP = FP + 1
    return [TN,FP, FN, TP]

In [136]:
print(confusion_matrix([1, 0, 1, 1], [0, 0, 1, 1])) #Expected to print: [1, 0, 1, 2]

[1, 0, 1, 2]


## Task : F
You need to complete the accuracy function partially defined in the file that takes a confusion matrix, i.e. the list of the four metrics: [TN, FP, FN, TP], in this order, and return accuracy. In case of Division by Zero error, return -1.

In [137]:
def accuracy(conf_mat):
    # Given a confusion matrix, i.e. the list of four metrics: [TN,FP, FN, TP], in this order
    # return accuracy

    
    acc_value = 0
    
    #@TODO
    #accuray = TN + TP / Total
    TN = conf_mat[0]
    TP = conf_mat[3]

    [TN, FP, FN, TP] = conf_mat
    correct = TN + TP
    total = TN + FP + FN + TP

    if total == 0: 
        return(-1)
        #acc_value = -1
    else:     
        acc_value = correct/total
    
    return acc_value

In [138]:
conf_mat = confusion_matrix([1, 0, 1, 1], [0, 0, 1, 1])  #Expected to print 0.75
print(accuracy(conf_mat))

0.75


## Task : G
You need to complete the precision function partially defined in the file that takes a confusion matrix, i.e. the list of the four metrics: [TN, FP, FN, TP], in this order, and return precision. It is also known as Positive Predictive Value (PPV). In case of Division by Zero error, return -1.

In [139]:
def precision(conf_mat):
    # Given a confusion matrix, i.e. the list of four metrics: [TN,FP, FN, TP], in this order
    # return precision. It is also known as Positive Predictive Value (PPV)
    [TN, FP, FN, TP] = conf_mat
    prec_value = 0
    #print("TN:",TN, "FP:",FP, "FN:",FN, "TP:", TP)
    
    #@TODO
    if (TP + FP) == 0:
        return(-1)
        #prec_value = -1
    else:    
        prec_value = TP/(TP+FP)

    return prec_value

In [140]:
conf_mat = confusion_matrix([1, 0, 1, 1], [0, 0, 1, 1])  #Expected to print 1.0
print(precision(conf_mat))

1.0


## Task : H
You need to complete the recall function partially defined in the file that takes a confusion matrix, i.e. the list of the four metrics: [TN, FP, FN, TP], in this order, and return recall. It is also known as Sensitivity, or True Positive Rate (TPR). In case of Division by Zero error, return -1.

In [141]:
def recall(conf_mat):
    # Given a confusion matrix, i.e. the list of four metrics: [TN,FP, FN, TP], in this order
    # return recall. It is also known as Sensitivity, or True Positive Rate (TPR)
    [TN, FP, FN, TP] = conf_mat
    rec_value = 0
    
    #@TODO
    if (TP + FN) == 0:
        return(-1)
        #rec_value = -1
    else:    
        rec_value = TP/(TP + FN)
    
    return rec_value

In [142]:
conf_mat = confusion_matrix([1, 0, 1, 1], [0, 0, 1, 1])  #Expected to print 0.6666666666
print(recall(conf_mat))

0.6666666666666666


## Task : I
You need to complete the F1 function partially defined in the file that takes a confusion matrix, i.e. the list of the four metrics: [TN, FP, FN, TP], in this order, and return F1. It is the harmonic mean of precision and recall. In case of Division by Zero error, return -1.

In [143]:
def F1(conf_mat):
    # Given a confusion matrix, i.e. the list of four metrics: [TN,FP, FN, TP], in this order
    # return F1 score. It is the harmonic mean of precision and recall
    [TN, FP, FN, TP] = conf_mat
    f1_value = 0


    #@TODO
    if (TP + FP) == 0:
        return(-1)
        #prec_value = -1
    else:     
        prec_value = TP/(TP+FP)
    
    if (TP +FN) == 0:
        return(-1)
        #rec_value = -1
    else: 
        rec_value = TP/(TP + FN)

    if (prec_value + rec_value) == 0:
        return(-1)
        #f1_value = -1
    else:
        f1_value = 2 * ((prec_value * rec_value)/(prec_value + rec_value))

    return f1_value

In [144]:
conf_mat = confusion_matrix([1, 0, 1, 1], [0, 0, 1, 1]) #Expected to print 0.8
print(F1(conf_mat))

0.8


## Task : J
You need to complete the MCC function defined in the file that takes a confusion matrix, i.e. the list of the four metrics: [TN, FP, FN, TP], in this order, and return Matthews Correlation Coefficient (MCC). In case of Division by Zero error, return -1.

In [145]:
import math
def MCC(conf_mat):
    # Given a confusion matrix, i.e. the list of four metrics: [TN,FP, FN, TP], in this order
    # return Matthews correlation coefficient (MCC)
    [TN, FP, FN, TP] = conf_mat
    mcc_value = 0
    
    #@TODO
    denominator = math.sqrt( ((TP + FP) * (TP + FN) * (TN + FP) * (TN + FN)) )
    if denominator == 0: 
        return(-1)
        #mcc_value = -1
    else:     
        mcc_value = ((TP * TN) - (FP * FN))/denominator 

    return mcc_value

In [146]:
conf_mat = confusion_matrix([1, 0, 1, 1], [0, 0, 1, 1]) #Expected to print 0.5773502691896258
print(MCC(conf_mat))

0.5773502691896258


## Task : K
You need to complete the FDR function defined in the file that takes a confusion matrix, i.e. the list of the four metrics: [TN, FP, FN, TP], in this order, and return False Discovery Rate (FDR). In case of Division by Zero error, return -1

In [147]:
def FDR(conf_mat):
    # Given a confusion matrix, i.e. the list of four metrics: [TN,FP, FN, TP], in this order
    # return False Discovery Rate (FDR)
    [TN, FP, FN, TP] = conf_mat
    fdr_value = 0
    
    #@TODO
    if (TP + FP) == 0: 
        return(-1)
        #fdr_value = -1
    else:     
        fdr_value = FP/(TP+FP)

    return fdr_value

In [148]:
conf_mat = confusion_matrix([1, 0, 1, 1], [0, 0, 1, 1]) #Expected to print 0.0
print(FDR(conf_mat))

0.0


## Task : L
Print as a dataframe containing:
 {model_name,acc,prec,rec,f1,mcc,FDR} for each of the N models (listed in models/*) after predicting the target variables of the validation data: "dataset/validation.csv"


In [149]:
#Print as a dataframe containing:
# {model_name,acc,prec,rec,f1,mcc,FDR} for each of the N models (listed in model_files) predicting the target variables
#  of the validation data.
import pickle
import numpy as np
import pandas as pd
import sys
import os
from tqdm import tqdm
from sklearn.linear_model import LogisticRegression
from sklearn.neural_network import MLPClassifier
from sklearn import tree
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
result = pd.DataFrame(columns=['model_name','Accuracy','Precision','Recall','F1','MCC','FDR'])

model_files=["models/Model_1.pkl","models/Model_2.pkl","models/Model_3.pkl",\
                            "models/Model_4.pkl","models/Model_5.pkl","models/Model_6.pkl"]


for file_name in tqdm(model_files):
    in_file = open(file_name,'rb')
    model = pickle.load(in_file)
    in_file.close()

    #Do the prediction by calling the "predict" member function of the model object on the
    # validation data with the 8 features
    X_test = validation.iloc[:,:-1]
    y_test = validation.iloc[:,-1]
    y_pred = model.predict(X_test)
    
    
    #@TODO: Complete and revise the following...
    #conf_mat = []
    conf_mat = confusion_matrix(y_test, y_pred)
    print(model)
    acc = 0
    acc = accuracy(conf_mat)
    
    prec = 0
    prec = precision(conf_mat)
    
    rec = 0
    rec = recall(conf_mat)
    
    f1 = 0
    f1 = F1(conf_mat)
    
    mcc = 0
    mcc = MCC(conf_mat)
    
    fdr = 0
    fdr = FDR(conf_mat)
    
    result = result.append(pd.Series([file_name,acc,prec,rec,f1,mcc,fdr],index=result.columns),ignore_index=True)


  0%|          | 0/6 [00:00<?, ?it/s]

LogisticRegression(random_state=1222)


 33%|███▎      | 2/6 [00:02<00:05,  1.25s/it]

KNeighborsClassifier(n_neighbors=3)


 50%|█████     | 3/6 [01:27<01:48, 36.15s/it]

SVC(gamma='auto')


 67%|██████▋   | 4/6 [01:27<00:45, 22.77s/it]

MLPClassifier(hidden_layer_sizes=(100, 100, 100), max_iter=500, random_state=21,
              solver='sgd', tol=1e-09)
DecisionTreeClassifier(max_depth=5)


100%|██████████| 6/6 [01:43<00:00, 17.22s/it]

SVC(C=0.025, kernel='linear')


In [150]:
print('Metrics after predicting the target variables of the validation data')
print(result)

Metrics after predicting the target variables of the validation data
           model_name  Accuracy  Precision   Recall        F1       MCC  \
0  models/Model_1.pkl  0.337583  0.334082   0.99995  0.500835  0.050929   
1  models/Model_2.pkl  0.334083  0.332910   0.99995  0.499518  0.028982   
2  models/Model_3.pkl  0.667667 -1.000000   0.00000 -1.000000 -1.000000   
3  models/Model_4.pkl  0.390483  0.352497   0.99659  0.520789  0.168805   
4  models/Model_5.pkl  0.333583  0.332744   0.99995  0.499330  0.024302   
5  models/Model_6.pkl  0.337500  0.334054   0.99995  0.500804  0.050516   

        FDR  
0  0.665918  
1  0.667090  
2 -1.000000  
3  0.647503  
4  0.667256  
5  0.665946  


## Task : M
Print the model name with path which is performing superior among the 6 pretrained models in terms of accuracy, given the performance result dataframe from “L”

In [151]:
result.loc[result['Accuracy'].idxmax()]

model_name    models/Model_3.pkl
Accuracy      0.667667          
Precision    -1.0               
Recall        0.0               
F1           -1.0               
MCC          -1.0               
FDR          -1.0               
Name: 2, dtype: object

## Task : N
Print the model name with path which is performing the worst among the 6 pretrained models in terms of recall, given the performance result dataframe from “L”

In [152]:
result.loc[result['Recall'].idxmin()]

model_name    models/Model_3.pkl
Accuracy      0.667667          
Precision    -1.0               
Recall        0.0               
F1           -1.0               
MCC          -1.0               
FDR          -1.0               
Name: 2, dtype: object

## Task : O
Scale all the features of the validation set using the formula, z = (x-m)/s,
    
    where m = mean of a feature in the training set: "dataset/training.csv"
    
    s = standard deviation of the feature in the training set: "dataset/training.csv"
    
    #  DO NOT SCALE the target feature.
    
    # At the end, return a tuple (X, y), with X being a numpy array of shape (N,8) and y is an N dim array and 

N is the total number of samples in the validation set: "dataset/validation.csv".

Store the scaled dataset in a variable (preferably a dataframe) named “X_validation_scaled”.


In [153]:
validation.head()

0    1         2         3         4         5         6         7  8
0 -315  75   0.205128  0.230769  0.307692  0.146667  0.280000  0.240000  0
1 -243  75   0.213333  0.186667  0.293333  0.400000  0.253333  0.186667  0
2 -289  77   0.192308  0.230769  0.282051  0.220779  0.220779  0.272727  0
3 -776  118  0.271186  0.203390  0.271186  0.200000  0.200000  0.291667  1
4 -388  77   0.226190  0.238095  0.250000  0.272727  0.272727  0.207792  1

In [154]:
X = validation.copy()
X.drop(8, axis=1, inplace=True)  

means = []
std_deviations = []

#TRAINING SET: calculate mean and standard deviation of each column, store in two lists
for feature in training.columns:
    if feature== 8:
        break;
    mean = training[feature].mean()
    s = training[feature].std()
    means.append(mean)
    std_deviations.append(s)    
#VALIDATION SET: scale all features but last (we dropped the target column at beginning)
for feature in validation:
    #b/c index error(validations set contains 1 more column, the target feature)
    if feature == 8:
        break; 
    X[feature] = (validation[feature] - means[feature])/std_deviations[feature]

#print(X.shape)

targets = validation[8]

X_validation_scaled = X
X_validation_scaled['targets'] = targets

print('Validation set scaled values: ')
print(X_validation_scaled)


Validation set scaled values: 
              0         1         2         3         4         5         6  \
0      0.775328 -1.002997 -0.325953  0.167255  1.069653 -1.676320  1.387452   
1      1.155149 -1.002997 -0.163911 -0.761209  0.723327  3.906838  0.758204   
2      0.912485 -0.910311 -0.579139  0.167255  0.451215 -0.042980 -0.009957   
3     -1.656587  0.989744  0.978644 -0.409145  0.189161 -0.500924 -0.500269   
4      0.390230 -0.910311  0.090006  0.321487 -0.321826  1.101893  1.215834   
...         ...       ...       ...       ...       ...       ...       ...   
59995 -1.292591  1.082430  0.231075 -0.129629  0.080167  0.233696  0.482927   
59996  0.305826  0.943401  0.518031  0.347192 -1.404098  0.091722 -0.064040   
59997 -2.031134  1.082430  0.231075 -0.305082  0.281151  0.233696 -0.303639   
59998  1.376712 -1.002997 -0.427228  0.080897  0.723327 -0.329198 -0.929301   
59999 -1.624935  1.082430  0.317418 -2.102599  0.765511 -0.317275 -0.106986   

              7  tar

## Task : P
Print as a dataframe containing:
  {model_name,acc,prec,rec,f1,mcc,FDR} for each of the N models (listed in model_files) after  predicting the target variables "y" (given) for "X" (the scaled validation dataset).


In [155]:
#@TODO
#Print as a dataframe containing:
# {model_name,acc,prec,rec,f1,mcc,FDR} for each of the N models (listed in model_files) predicting the target variables
#  of the validation data.
import pickle
import numpy as np
import pandas as pd
import sys
import os
from tqdm import tqdm
from sklearn.linear_model import LogisticRegression
from sklearn.neural_network import MLPClassifier
from sklearn import tree
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
result_scaled = pd.DataFrame(columns=['model_name','Accuracy','Precision','Recall','F1','MCC','FDR'])

model_files=["models/Model_1.pkl","models/Model_2.pkl","models/Model_3.pkl",\
                            "models/Model_4.pkl","models/Model_5.pkl","models/Model_6.pkl"]


for file_name in tqdm(model_files):
    in_file = open(file_name,'rb')
    model = pickle.load(in_file)
    in_file.close()

    #Do the prediction by calling the "predict" member function of the model object on the
    # validation data with the 8 features
    X_test = X_validation_scaled.iloc[:,:-1]
    y_test = X_validation_scaled.iloc[:,-1]
    y_pred = model.predict(X_test)
    
    
    #@TODO: Complete and revise the following...
    conf_mat = []
    conf_mat = confusion_matrix(y_test, y_pred)
    print(model)
    print(conf_mat)
    acc = 0
    acc = accuracy(conf_mat)
    
    prec = 0
    prec = precision(conf_mat)
    
    rec = 0
    rec = recall(conf_mat)
    
    f1 = 0
    f1 = F1(conf_mat)
    
    mcc = 0
    mcc = MCC(conf_mat)
    
    fdr = 0
    fdr = FDR(conf_mat)
    
    result_scaled = result_scaled.append(pd.Series([file_name,acc,prec,rec,f1,mcc,fdr],index=result_scaled.columns),ignore_index=True)


  0%|          | 0/6 [00:00<?, ?it/s]

LogisticRegression(random_state=1222)
[38993, 1067, 1859, 18081]


 33%|███▎      | 2/6 [00:05<00:11,  2.81s/it]

KNeighborsClassifier(n_neighbors=3)
[38840, 1220, 1918, 18022]


 50%|█████     | 3/6 [01:05<01:19, 26.50s/it]

SVC(gamma='auto')
[38990, 1070, 1013, 18927]


 67%|██████▋   | 4/6 [01:05<00:33, 16.74s/it]

MLPClassifier(hidden_layer_sizes=(100, 100, 100), max_iter=500, random_state=21,
              solver='sgd', tol=1e-09)
[39041, 1019, 1015, 18925]
DecisionTreeClassifier(max_depth=5)
[37920, 2140, 3012, 16928]


100%|██████████| 6/6 [01:21<00:00, 13.51s/it]

SVC(C=0.025, kernel='linear')
[38818, 1242, 1525, 18415]


In [156]:
print('Metrics of Models after predicting using scaled values: ')
print(result_scaled)

Metrics of Models after predicting using scaled values: 
           model_name  Accuracy  Precision    Recall        F1       MCC  \
0  models/Model_1.pkl  0.951233  0.944276   0.906770  0.925143  0.889404   
1  models/Model_2.pkl  0.947700  0.936597   0.903811  0.919912  0.881411   
2  models/Model_3.pkl  0.965283  0.946492   0.949198  0.947843  0.921828   
3  models/Model_4.pkl  0.966100  0.948907   0.949097  0.949002  0.923614   
4  models/Model_5.pkl  0.914133  0.887770   0.848947  0.867925  0.804802   
5  models/Model_6.pkl  0.953883  0.936816   0.923521  0.930121  0.895760   

        FDR  
0  0.055724  
1  0.063403  
2  0.053508  
3  0.051093  
4  0.112230  
5  0.063184  


## Task : Q
Print the model name with path which is performing superior in terms of accuracy, given the performance result dataframe from “P”

In [157]:
result_scaled.loc[result['Accuracy'].idxmax()]

model_name    models/Model_3.pkl
Accuracy      0.965283          
Precision     0.946492          
Recall        0.949198          
F1            0.947843          
MCC           0.921828          
FDR           0.053508          
Name: 2, dtype: object

## Task : R
Print the model name with path which is performing the worst in terms of recall, given the performance result dataframe from “P”

In [158]:
result_scaled.loc[result['Recall'].idxmin()]

model_name    models/Model_3.pkl
Accuracy      0.965283          
Precision     0.946492          
Recall        0.949198          
F1            0.947843          
MCC           0.921828          
FDR           0.053508          
Name: 2, dtype: object

## Task : S
Flip the prediction of Model 1, and then compute and
 print as a dataframe containing: {acc,prec,rec,f1,mcc,FDR} 
on the original (i.e., not-scaled) validation dataset: "dataset/validation.csv"


In [159]:
model_1_validation = validation.copy()

#flip predictions 
flipped_array = []

for item in model_1_validation[8]:
    if item == 0:
        item = 1
        flipped_array.append(item)
    else: 
        item = 0
        flipped_array.append(item)

model_1_validation[8] = flipped_array
print("Flipped predictions for Model 1")
print(model_1_validation) 

#@TODO
#Print as a dataframe containing:
# {model_name,acc,prec,rec,f1,mcc,FDR} for each of the N models (listed in model_files) predicting the target variables
#  of the validation data.
import pickle
import numpy as np
import pandas as pd
import sys
import os
from tqdm import tqdm
from sklearn.linear_model import LogisticRegression
from sklearn.neural_network import MLPClassifier
from sklearn import tree
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
model_1 = pd.DataFrame(columns=['model_name','Accuracy','Precision','Recall','F1','MCC','FDR'])

model_files=["models/Model_1.pkl"]

for file_name in tqdm(model_files):
    in_file = open(file_name,'rb')
    model = pickle.load(in_file)
    in_file.close()

    #Do the prediction by calling the "predict" member function of the model object on the
    # validation data with the 8 features
    X_test = model_1_validation.iloc[:,:-1]
    y_test = model_1_validation.iloc[:,-1]
    y_pred = model.predict(X_test)
    
    
    #@TODO: Complete and revise the following...
    #conf_mat = []
    conf_mat = confusion_matrix(y_test, y_pred)
    print(model)
    acc = 0
    acc = accuracy(conf_mat)
    
    prec = 0
    prec = precision(conf_mat)
    
    rec = 0
    rec = recall(conf_mat)
    
    f1 = 0
    f1 = F1(conf_mat)
    
    mcc = 0
    mcc = MCC(conf_mat)
    
    fdr = 0
    fdr = FDR(conf_mat)
    
    model_1 = model_1.append(pd.Series([file_name,acc,prec,rec,f1,mcc,fdr],index=model_1.columns),ignore_index=True)


Flipped predictions for Model 1
         0    1         2         3         4         5         6         7  8
0     -315  75   0.205128  0.230769  0.307692  0.146667  0.280000  0.240000  1
1     -243  75   0.213333  0.186667  0.293333  0.400000  0.253333  0.186667  1
2     -289  77   0.192308  0.230769  0.282051  0.220779  0.220779  0.272727  1
3     -776  118  0.271186  0.203390  0.271186  0.200000  0.200000  0.291667  0
4     -388  77   0.226190  0.238095  0.250000  0.272727  0.272727  0.207792  0
...    ...  ..        ...       ...       ...       ...       ...       ... ..
59995 -707  120  0.233333  0.216667  0.266667  0.233333  0.241667  0.250000  0
59996 -404  117  0.247863  0.239316  0.205128  0.226891  0.218487  0.252101  1
59997 -847  120  0.233333  0.208333  0.275000  0.233333  0.208333  0.258333  0
59998 -201  75   0.200000  0.226667  0.293333  0.207792  0.181818  0.298701  1
59999 -770  120  0.237705  0.122951  0.295082  0.208333  0.216667  0.266667  0

[60000 rows x 9 col

100%|██████████| 1/1 [00:00<00:00, 17.00it/s]

LogisticRegression(random_state=1222)


In [160]:
print('model 1 metrics after flipping predictions:')
print(model_1)

model 1 metrics after flipping predictions:
           model_name  Accuracy  Precision    Recall        F1       MCC  \
0  models/Model_1.pkl  0.662417  0.665918   0.992112  0.796928 -0.050929   

        FDR  
0  0.334082  


## Task : T
Say, in a confusion matrix, the values of the four metrics are: TP=90, TN=1, FP=4, FN=5. Compute F1_original and MCC_original denoting the F1 and MCC scores.

Now, flip the predictions, i.e., positives are now will be predicted as negative, and negatives are going to be predicted as positive. 

Then, compute F1_flipped and MCC_flipped, denoting corresponding F1 and MCC scores. 

Print/Return the new {TP, TN, FP, FN, F1_original,MCC_original,F1_flipped, MCC_flipped, COMMENT_string} as a dataframe, where COMMENT_string is a string that will be no longer than 200 characters but is going to be your comment about the F1 and MCC values for the two cases.


In [161]:
import pandas as pd
pd.set_option('display.max_colwidth', -1)  # or 199

#[TN, FP, FN, TP] = conf_mat
conf_matrix = [1, 4, 5, 90]

F1_original = F1(conf_mat)
MCC_original = MCC(conf_mat)

print("F1_original: ", F1_original, "MCC_original:", MCC_original)

F1_original:  0.7969281052304422 MCC_original: -0.050929488373451064


C:\Users\odyss\AppData\Local\Temp/ipykernel_2052/4039341406.py:2: FutureWarning: Passing a negative integer is deprecated in version 1.0 and will not be supported in future version. Instead, use None to not limit the column width.
  pd.set_option('display.max_colwidth', -1)  # or 199


In [162]:
TP = 90
TN = 1 
FP = 4 
FN = 5 

#flip prdeictions
TN_flipped = 4 #FP
FN_flipped = 90 #TP
FP_flipped = 1 #TN
TP_flipped = 5 #FN

conf_matrix = [90,5,4,1]
F1_flipped = F1(conf_mat)
MCC_flipped = MCC(conf_mat)


df = [TP, TN, FP, FN, F1_original,MCC_original, F1_original,MCC_original]
print(df)
print( " The F1 and MCC values for the two cases are the same.")

[90, 1, 4, 5, 0.7969281052304422, -0.050929488373451064, 0.7969281052304422, -0.050929488373451064]
 The F1 and MCC values for the two cases are the same.


## For Graduate Students only

## Task: U
This task is a follow up of Task P. There is always cost associated with misclassifications. For instance, if a model predicts a ncRNA (class=1) to be non ncRNA (class=0), further verfication will then follow that includes going through the next generation sequencing of those samples costing USD 20 per sample. On the other hand, cost of predicting a non ncRNA to be ncRNA insignificant as most researchers do not care for ncRNAs. They might put it in a piece of paper as a note costing USD 1 per 5 samples. The same cost applies to correct predictions too.

What is the cost of predictions with each of the six models?

In [163]:
#cost matrix 
#cost of FP : $20/sample 

#cost of FN, TP, TN : $0.20/samples


In [164]:
def cost_matrix(conf_mat):
    [TN, FP, FN, TP] = conf_mat
    #fdr_value = 0
    
#use Confusion Matrix variables to calcualte cost         
    FN_cost = 20
    FP_cost = 0.2
    TP_cost = 0.2
    TN_cost = 0.2
    
    TP = TP*(0.2)
    FN = FN*(20)
    FP = FP*(0.2)
    TN = TN*(0.2)
    total_cost = TP+FP+FN+TN    
    return total_cost

In [165]:
#@TODO
#Print as a dataframe containing:
# {model_name,acc,prec,rec,f1,mcc,FDR} for each of the N models (listed in model_files) predicting the target variables
#  of the validation data.
import pickle
import numpy as np
import pandas as pd
import sys
import os
from tqdm import tqdm
from sklearn.linear_model import LogisticRegression
from sklearn.neural_network import MLPClassifier
from sklearn import tree
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
result_cost = pd.DataFrame(columns=['model_name','Accuracy','Precision','Recall','F1','MCC','FDR','Cost'])

model_files=["models/Model_1.pkl","models/Model_2.pkl","models/Model_3.pkl",\
                            "models/Model_4.pkl","models/Model_5.pkl","models/Model_6.pkl"]


for file_name in tqdm(model_files):
    in_file = open(file_name,'rb')
    model = pickle.load(in_file)
    in_file.close()

    #Do the prediction by calling the "predict" member function of the model object on the
    # validation data with the 8 features
    X_test = X_validation_scaled.iloc[:,:-1]
    y_test = X_validation_scaled.iloc[:,-1]
    y_pred = model.predict(X_test)
    
    
   #@TODO: Complete and revise the following...
    #conf_mat = []
    conf_mat = confusion_matrix(y_test, y_pred)
    print(model)

    cost = 0
    cost = cost_matrix(conf_mat)
    
    acc = 0
    acc = accuracy(conf_mat)
    
    prec = 0
    prec = precision(conf_mat)
    
    rec = 0
    rec = recall(conf_mat)
    
    f1 = 0
    f1 = F1(conf_mat)
    
    mcc = 0
    mcc = MCC(conf_mat)
    
    fdr = 0
    fdr = FDR(conf_mat)
    
    result_cost = result_cost.append(pd.Series([file_name,acc,prec,rec,f1,mcc,fdr,cost],index=result_cost.columns),ignore_index=True)
    


  0%|          | 0/6 [00:00<?, ?it/s]

LogisticRegression(random_state=1222)


 33%|███▎      | 2/6 [00:05<00:10,  2.71s/it]

KNeighborsClassifier(n_neighbors=3)


 50%|█████     | 3/6 [01:21<01:40, 33.44s/it]

SVC(gamma='auto')


 67%|██████▋   | 4/6 [01:22<00:42, 21.10s/it]

MLPClassifier(hidden_layer_sizes=(100, 100, 100), max_iter=500, random_state=21,
              solver='sgd', tol=1e-09)
DecisionTreeClassifier(max_depth=5)


100%|██████████| 6/6 [01:37<00:00, 16.30s/it]

SVC(C=0.025, kernel='linear')


In [166]:
print('Model metrics including cost:')
result_cost

Model metrics including cost:


model_name  Accuracy  Precision    Recall        F1       MCC  \
0  models/Model_1.pkl  0.951233  0.944276   0.906770  0.925143  0.889404   
1  models/Model_2.pkl  0.947700  0.936597   0.903811  0.919912  0.881411   
2  models/Model_3.pkl  0.965283  0.946492   0.949198  0.947843  0.921828   
3  models/Model_4.pkl  0.966100  0.948907   0.949097  0.949002  0.923614   
4  models/Model_5.pkl  0.914133  0.887770   0.848947  0.867925  0.804802   
5  models/Model_6.pkl  0.953883  0.936816   0.923521  0.930121  0.895760   

        FDR     Cost  
0  0.055724  48808.2  
1  0.063403  49976.4  
2  0.053508  32057.4  
3  0.051093  32097.0  
4  0.112230  71637.6  
5  0.063184  42195.0

## Task: V
Please comment on which of the six models is the best on the cost basis.

Model_4 has the lowest cost 

## Task: W
Please comment on which of the six models is the best overall. Explain your answer.

All the models except model_5 have similar Accuracy, Precision, Recall, F1, MCC, FDR, and cost. Model_5 performd worse in all categories. Out of the Models 1 - 4, Model 4 has the least cost, and with the other metrics being comparable, Model_4 performed best based on the lowest cost. 

In [167]:
#see markdown answer above

## Task: X
Again a followup of Tasks O and P: Please scale the given validation dataset with an alternate scaling technique you can think of and repeat Task P with the modified scaled validation.

In [168]:
# Scale validation data by Mean-Max Normilization: X(new) = Xi – min(X) / max(x) – min(x)
X = validation.copy()
X.drop(8, axis=1, inplace=True)  


mins = []
maxs = []

#TRAINING SET: calculate min and max of each column, store in two lists
for feature in training.columns:
    if feature== 8:
        break;
    
    maxVal = training.loc[training[feature].idxmax()][feature]
    maxs.append(maxVal)
                          
    minVal = training.loc[training[feature].idxmin()][feature]
    mins.append(minVal) 
#print('max: ', maxs)
#print('min:', mins)
#VALIDATION SET: scale all features but last (we dropped the target column at beginning)
for feature in validation:
    #b/c index error(validations set contains 1 more column, the target feature)
    if feature == 8:
        break;
        
    else: 
        X[feature] = (validation[feature] - mins[feature])/(maxs[feature] - mins[feature])

#print(X.shape)

targets = validation[8]

min_max_scaled = X
min_max_scaled['targets'] = targets

print("Scaled Min_Max Validation values: ")
print(min_max_scaled)


Scaled Min_Max Validation values: 
              0         1         2         3         4         5         6  \
0      0.831370  0.121212  0.267297  0.393723  0.806452  0.110821  0.551213   
1      0.869914  0.121212  0.289259  0.252641  0.754839  0.788889  0.465905   
2      0.845289  0.151515  0.232984  0.393723  0.714286  0.309189  0.361765   
3      0.584582  0.772727  0.444107  0.306138  0.675232  0.253572  0.295293   
4      0.792291  0.151515  0.323672  0.417159  0.599079  0.448232  0.527946   
...         ...       ...       ...       ...       ...       ...       ...   
59995  0.621520  0.803030  0.342791  0.348611  0.658989  0.342791  0.428586   
59996  0.783726  0.757576  0.381681  0.421065  0.437788  0.325548  0.354433   
59997  0.546574  0.803030  0.342791  0.321950  0.688941  0.342791  0.321950   
59998  0.892398  0.121212  0.253572  0.380601  0.754839  0.274428  0.237129   
59999  0.587794  0.803030  0.354493  0.048813  0.761126  0.275876  0.348611   

              7 

In [169]:
#@TODO
#Print as a dataframe containing:
# {model_name,acc,prec,rec,f1,mcc,FDR} for each of the N models (listed in model_files) predicting the target variables
#  of the validation data.
import pickle
import numpy as np
import pandas as pd
import sys
import os
from tqdm import tqdm
from sklearn.linear_model import LogisticRegression
from sklearn.neural_network import MLPClassifier
from sklearn import tree
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
result_scaled_min_max = pd.DataFrame(columns=['model_name','Accuracy','Precision','Recall','F1','MCC','FDR'])

model_files=["models/Model_1.pkl","models/Model_2.pkl","models/Model_3.pkl",\
                            "models/Model_4.pkl","models/Model_5.pkl","models/Model_6.pkl"]


for file_name in tqdm(model_files):
    in_file = open(file_name,'rb')
    model = pickle.load(in_file)
    in_file.close()

    #Do the prediction by calling the "predict" member function of the model object on the
    # validation data with the 8 features
    X_test = min_max_scaled.iloc[:,:-1]
    y_test = min_max_scaled.iloc[:,-1]
    y_pred = model.predict(X_test)
    
    
    #@TODO: Complete and revise the following...
    conf_mat = []
    conf_mat = confusion_matrix(y_test, y_pred)
    print(model)
    #print(y_test, 'predicted targets: ', y_pred)
    acc = 0
    acc = accuracy(conf_mat)
    
    prec = 0
    prec = precision(conf_mat)
    
    rec = 0
    rec = recall(conf_mat)
    
    f1 = 0
    f1 = F1(conf_mat)
    
    mcc = 0
    mcc = MCC(conf_mat)
    
    fdr = 0
    fdr = FDR(conf_mat)
    
    result_scaled_min_max = result_scaled_min_max.append(pd.Series([file_name,acc,prec,rec,f1,mcc,fdr],index=result_scaled_min_max.columns),ignore_index=True)
    

  0%|          | 0/6 [00:00<?, ?it/s]

LogisticRegression(random_state=1222)


 33%|███▎      | 2/6 [00:23<00:47, 11.80s/it]

KNeighborsClassifier(n_neighbors=3)


 50%|█████     | 3/6 [01:22<01:34, 31.39s/it]

SVC(gamma='auto')


 67%|██████▋   | 4/6 [01:22<00:39, 19.78s/it]

MLPClassifier(hidden_layer_sizes=(100, 100, 100), max_iter=500, random_state=21,
              solver='sgd', tol=1e-09)
DecisionTreeClassifier(max_depth=5)


100%|██████████| 6/6 [01:38<00:00, 16.34s/it]

SVC(C=0.025, kernel='linear')


In [170]:
print("Model metrics after scaling based on Min-Max Method")
print(result_scaled_min_max)
print('metrics are similar because divison of zero is returning the error (-1) value for Precision')

Model metrics after scaling based on Min-Max Method
           model_name  Accuracy Precision   Recall        F1       MCC  FDR
0  models/Model_1.pkl  0.667667  -1        0.00000  -1        -1        -1 
1  models/Model_2.pkl  0.667667  -1        0.00000  -1        -1        -1 
2  models/Model_3.pkl  0.667667  -1        0.00000  -1        -1        -1 
3  models/Model_4.pkl  0.667717  1.0       0.00015  0.000301  0.010023  0.0
4  models/Model_5.pkl  0.667667  -1        0.00000  -1        -1        -1 
5  models/Model_6.pkl  0.667667  -1        0.00000  -1        -1        -1 
metrics are similar because divison of zero is returning the error (-1) value for Precision


## Task: Y
Print the model name with path which is performing superior in terms of accuracy, given the performance result dataframe from “X”

In [171]:
result_scaled_min_max.loc[result['Accuracy'].idxmax()]

model_name    models/Model_3.pkl
Accuracy      0.667667          
Precision     -1                
Recall        0.0               
F1            -1                
MCC           -1                
FDR           -1                
Name: 2, dtype: object

## Task: Z
Print the model name with path which is performing the worst in terms of recall, given the performance result dataframe from “X”

In [172]:
result_scaled_min_max.loc[result['Recall'].idxmin()]

model_name    models/Model_3.pkl
Accuracy      0.667667          
Precision     -1                
Recall        0.0               
F1            -1                
MCC           -1                
FDR           -1                
Name: 2, dtype: object

I used the Min-Max equation to scale the values for the validation set. This turned out to be a poor method for scaling the values, because it predicted every sample to be a negative. Which ultimately gave FP & TP values of 0, and avg accuracy accross all models was 67%, which is lower than when we scaled using the Stndardization method above. 

Overall this method of scaling was not good 